### ISSUE TITLE
Unable to overwrite worker's onmessage when using MODULARIZE

### ISSUE THREAD

**Issue by miloszmaki** [emscripten-core/emscripten#20192]
Consider this example, which creates a new worker thread. The worker overwrites its `onmessage` to handle custom messages. Then the main thread sends a custom message to the worker.

```cpp
#include <cstdio>
#include <emscripten.h>
#include <thread>

int main() {
  printf("hello from main\n");

  std::thread t{[](){
    EM_ASM({
        console.log("hello from thread");
        const def_onmsg = self.onmessage;
        self.onmessage = (e) => {
                console.log("handling message:", e, e.data, e.data.cmd);
                if (e["data"]["cmd"] != "custom1" && e["data"]["cmd"] != "custom2") {
                        def_onmsg(e);
                }
        };
        // self.onmessage({"data": {"cmd": "custom1", "id": 1}}); // this works always
    });
  }};

  EM_ASM({
  // setTimeout(()=>{ // doesn't help either
    const threads = Module["PThread"].pthreads;
    const id = Object.keys(threads)[0];
    const worker = threads[id];
    worker.postMessage({"cmd": "custom2", "id": 2}); // this fails with MODULARIZE
  // },1000);
  });

  return 0;
}
```

It works properly ("custom2" message gets handled by the worker) when I build with:
`emcc main.cpp -o test.html --std=c++20 -pthread -s PTHREAD_POOL_SIZE_STRICT=0`

However, it stops working (the console prints error "worker.js received unknown command custom2") when I build with:
`emcc main.cpp -o test.html --std=c++20 -pthread -s PTHREAD_POOL_SIZE_STRICT=0 -s MODULARIZE=1 -s EXPORT_ES6=1`

My output of `emcc -v` is:
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.45 (ef3e4e3b044de98e1811546e0bc605c65d3412f4)
clang version 18.0.0 (https://github.com/llvm/llvm-project d1e685df45dc5944b43d2547d0138cd4a3ee4efe)
Target: wasm32-unknown-emscripten
Thread model: posix
```

Interestingly, it was working with some older versions (at least I checked `3.1.15`).

**Comment by sbc100** [comment]
I'm not sure we have ever officially support doing this kind of thing.  We have had others ask about it in the past though so there are at least few folks who want to do it.  If we do want to support it we should add some tests that do this so it doesn't break again. 

We should also make some kind of official API for going from a pthread ID to a worker handle in JS I suppose?

**Comment by miloszmaki** [comment]
I see. That would be great if you decide to support this. I think that being able to establish the communication between workers is quite important. Do you know of any workarounds available currently? One I can think of, is to call some C++ function from JS and then do the thread communication on the C++ side. Do you think it's a good way to go?

**Comment by sbc100** [comment]
> I see. That would be great if you decide to support this. I think that being able to establish the communication between workers is quite important. Do you know of any workarounds available currently? One I can think of, is to call some C++ function from JS and then do the thread communication on the C++ side. Do you think it's a good way to go?

Normally pthread communication happens through shared memory.  is there some reason you can't do that in this case?  The fact that pthreads are implemented as web workers is kind of an implementation detail that ideally you would not rely on.  Is there some reason you need to use postMessage rather than shared memory?

**Comment by sbc100** [comment]
> do the thread communication on the C++ side.

Yes, I would recommend doing your thread communication using the pthread APIs (or C++ thread APIs) and shared memory.

**Comment by miloszmaki** [comment]
> Is there some reason you need to use postMessage rather than shared memory?
I'm capturing frames from the camera on the main thread and need to transfer the ownership of a [`VideoFrame`](https://developer.mozilla.org/en-US/docs/Web/API/VideoFrame) to the worker thread, which handles the frame by processing its data and eventually calling `close()` on it. Not sure if this is possible using shared memory only.

**Comment by sbc100** [comment]
Yes, I can see that in that case you would need to use postMessage.

@juj doesn't this sounds like a reasonable use case for postMessage or is there some way this can be done using offscreen canvas perhaps?

**Comment by sbc100** [comment]
Wait, if you are processing the data using native code then don't you have to copy the data out of the VideoFrame and into linear memory anyway?  Why not copy the data out on the main thread?

**Comment by miloszmaki** [comment]
Right, I'm copying the data for further processing in C++. However, I'm concerned about the efficiency and would like to offload work from the main thread if possible. I think I will need to try and profile the approach you suggested to know if it works for me.

**Comment by juj** [comment]
> I'm not sure we have ever officially support doing this kind of thing.

I think we should support this. It helps composability with other libraries, and helps users utilize the existing constructs that they might have and know about. Alternatively, we should clearly document that the .onmessage parameter is off limits to end users.

We might code golf the size a bit for this in the future to use .onmessage, but I think the better default is to use addEventListener. There are too many tutorials and other articles that showcase modifying .onmessage that it's better to let users have it.

**Comment by miloszmaki** [comment]
> Why not copy the data out on the main thread?

I tried this approach suggested by @sbc100 and it turns out to work quite well.

**Comment by miloszmaki** [comment]
However, I'm still a bit concerned with this new approach, since I need to do some kind of synchronization (e.g. mutex) which requires blocking on the main thread (not recommended, as pointed out in the [documentation](https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread)).

**Comment by miloszmaki** [comment]
@sbc100 @juj any updates on that? Is it possible to merge https://github.com/emscripten-core/emscripten/pull/20235 ?

**Comment by sbc100** [comment]
I'd be OK with landing #20235 but I think we should add a test/example and some docs about how to do it in the blessed way.

**Comment by juj** [comment]
Sorry, I never quite had time to design the tests for that PR. I'll see if I can get a bit of time slotted for this soon.

**Comment by barathrm** [comment]
I think I hit the same issue, but I'm new to emscripten and web workers etc so I may be off.

I'm building C++ code to js and wasm and then manually "spawning" the runtime inside a worker created with `new Worker("worker.js")` where `worker.js` will do `self.importScripts('myemscriptenstuff.js');`. 

I need to communicate with the C++ library wrapped in emscripten from outside the worker, so I tried configuring my own `onmessage` to be able to post messages to it and trigger usage of the library, but then I noticed that this breaks spawning pthreads.

For those coming after me, my work-around is to simply forward `onmessage` events I dont care about for my own uses to the runtime's `self.handleMessage()` function, which seems to do the trick.
